### PR TITLE
Increase ARC meta data limits

### PR DESCRIPTION
**Pull request by behlendorf**
### Motivation and Context

Issue #9966

### Description

Historically the ARC has limited meta data usage to some fraction
of the total ARC size.  The rational behind this was to ensure that
the more important metadata blocks wouldn't result in all data blocks
(even frequently used ones) being evicted from the cache.

However, this concern doesn't appear to have been entirely warranted.
For meta data heavy workloads we want the entire ARC to be available
for these blocks to improve the cache hit rate.  Furthermore, in
practice the most important data blocks will still remain cached as
long as they are frequently accessed.

For reasons very similar to the above a 10% limit was also imposed
on the faction of the ARC which can be consumed by dnodes.  For many
workloads this is fine.  However, for metadata heavy workloads it can
result in a lot of needless churn in the ARC and wasted CPU time spent
evicting dnodes to honor this limit.  To avoid this it's reasonable
to allow dnodes to take a larger fraction of the ARC when needed and
then be evicted by the VFS which can do this more efficiently.

This change increases the zfs_arc_meta_limit_percent to 100% of the
ARC, and the zfs_arc_dnode_limit_percent to 75% of the meta limit.
### How Has This Been Tested?

Locally compiled.  Updated values selected based on feedback from
previously filed issues and previous experience manually tuning ZFS
for specific workloads.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Review by adamdmoss — APPROVED** [review]

**Review by ahrens — APPROVED** [review]

**Comment by adamdmoss** [comment]
I have a fs-crawling backup script which ran in 95sec before this change, reliably 67sec after.

(The timing is from the second of two back-to-back crawls, thus with a warmed ARC - which isn't typical for a daily backup - but still, clearly the change is doing something positive for this sort of workload. :grinning:  )

**Comment by behlendorf** [comment]
@adamdmoss that's great, thanks for testing with your workload!  So my one concern with this change is that it _might_ take a while for the VFS to decide to drop all/most those dnodes from the cache in order to make room for something else.  Could you try performing some normal large file IO after your fs-crawling script and check how the ARC adapts to the new workload?

**Review by gmelikov — APPROVED** [review]

**Comment by adamdmoss** [comment]
That appears to be a valid concern.

arc_summary3's reported DNODE cache size raises to 1.2GB and I couldn't seem to dislodge that short of using /proc/sys/vm/drop_caches.

A bit more worrying, the META field raised to 5.3GB and again no amount of non-meta reads appears to lower that.  Weirder, a repeated smaller simpler meta-heavy workload after that other crawl never stopped churned the disk heavily, plateauing in a mediocre disk-churning place.  (Again, using /proc/sys/vm/drop_caches made everything good again.)

A 5GB non-metadata read was getting somewhat cached (about 1.3GB on top of the apparently-pinned 5.3GB meta), but plateauing at a mediocre place.  After /proc/sys/vm/drop_caches it was trivially entirely cached.

Haven't gone back to double-check that this differs from pre-PR behavior (perhaps tonight).

**Comment by adamdmoss** [comment]
(I have some patches kicking around from a while back where I 'fixed' some metadata eviction ARC issues but I just shelved them because they didn't solve the [OOM/exhaustion] issues I was actually chasing at the time; I'll give those a spin in combo with this PR just to see if they help.)

**Comment by adamdmoss** [comment]
Benchmarking master vs PR (in two columns to hopefully ease comparison):
```
master@f8ce8aed                                                         master+PR
Starting tests...							Starting tests...
Big metadata crawl:	1:36.43 real,	10.86 user,	46.02 sys	Big metadata crawl:	1:42.50 real,	11.54 user,	46.71 sys
Big metadata crawl:	1:18.05 real,	11.63 user,	29.50 sys	Big metadata crawl:	1:09.98 real,	10.97 user,	26.25 sys
Little metadata crawl:	0:16.70 real,	0.44 user,	7.71 sys	Little metadata crawl:	0:17.66 real,	0.54 user,	8.27 sys
Little metadata crawl:	0:01.88 real,	0.27 user,	1.51 sys	Little metadata crawl:	0:05.50 real,	0.44 user,	3.12 sys
Little metadata crawl:	0:01.91 real,	0.24 user,	1.57 sys	Little metadata crawl:	0:05.53 real,	0.44 user,	3.14 sys
Little metadata crawl:	0:01.90 real,	0.25 user,	1.55 sys	Little metadata crawl:	0:05.49 real,	0.38 user,	3.14 sys
Assorted 5GB data read:	0:05.25 real,	0.60 user,	1.29 sys	Assorted 5GB data read:	0:07.14 real,	0.55 user,	1.54 sys
Assorted 5GB data read:	0:04.90 real,	0.53 user,	1.39 sys	Assorted 5GB data read:	0:07.11 real,	0.60 user,	1.54 sys
Assorted 5GB data read:	0:04.90 real,	0.58 user,	1.42 sys	Assorted 5GB data read:	0:07.10 real,	0.56 user,	1.57 sys
Assorted 5GB data read:	0:04.88 real,	0.53 user,	1.39 sys	Assorted 5GB data read:	0:07.11 real,	0.53 user,	1.63 sys
Assorted 5GB data read:	0:04.90 real,	0.56 user,	1.43 sys	Assorted 5GB data read:	0:07.08 real,	0.56 user,	1.56 sys
DROPPED CACHES...							DROPPED CACHES...
Little metadata crawl:	0:09.03 real,	0.35 user,	4.28 sys	Little metadata crawl:	0:07.13 real,	0.36 user,	4.15 sys
Little metadata crawl:	0:01.52 real,	0.20 user,	1.35 sys	Little metadata crawl:	0:01.53 real,	0.23 user,	1.33 sys
Little metadata crawl:	0:01.52 real,	0.24 user,	1.31 sys	Little metadata crawl:	0:01.53 real,	0.23 user,	1.33 sys
Little metadata crawl:	0:01.51 real,	0.24 user,	1.29 sys	Little metadata crawl:	0:01.53 real,	0.22 user,	1.35 sys
Assorted 5GB data read:	0:05.18 real,	0.55 user,	1.46 sys	Assorted 5GB data read:	0:04.94 real,	0.61 user,	1.35 sys
Assorted 5GB data read:	0:00.57 real,	0.27 user,	1.45 sys	Assorted 5GB data read:	0:00.56 real,	0.34 user,	1.37 sys
Assorted 5GB data read:	0:00.56 real,	0.29 user,	1.43 sys	Assorted 5GB data read:	0:00.57 real,	0.28 user,	1.45 sys
Assorted 5GB data read:	0:00.58 real,	0.31 user,	1.36 sys	Assorted 5GB data read:	0:00.56 real,	0.30 user,	1.43 sys
Assorted 5GB data read:	0:00.56 real,	0.30 user,	1.42 sys	Assorted 5GB data read:	0:00.57 real,	0.28 user,	1.43 sys
```

**Comment by adamdmoss** [comment]
Not sure exactly what to conclude from those results apart from: metadata/dnodes do look quite 'sticky', so raising the limit on these does appear to pin even more data for longer (perhaps almost indefinitely).

Attached is the expanded data which includes a basic ARC breakdown courtesy of `arc_summary3`
[new.txt](https://github.com/openzfs/zfs/files/5948617/new.txt)
[old.txt](https://github.com/openzfs/zfs/files/5948618/old.txt)

**Comment by adamdmoss** [comment]
Okay, I lowered my arc_max from 16GB to 6GB, so the test case's dnode use is above the dnode limit in master and below the dnode limit in this PR.
I thought it'd make the difference in the PR more obvious, but this situation shows no practical difference in performance between master and the PR.

(The big metadata crawl's dnode usage busts way through the dnode limit on master, only lowering lowering back down to the limit when it comes to repeatedly loading the 5GB data.  In the PR it the dnode usage is the same, except it doesn't lower again, already being below the dnode limit there.  Either way, no practical performance difference.)

So that's kind of a wash I guess - in the lower arc_max case there's no cost or benefit to this PR.  With the original larger arc_max, the higher dnode limit appears to also make metadata stickier (and apparently harder to displace by data or even by more MFU metadata), which is pretty good for the successive large metadata crawls but bad for everything else.

**Comment by adamdmoss** [comment]
> So my one concern with this change is that it might take a while for the VFS to decide to drop all/most those dnodes from the cache in order to make room for something else.

TL;DR: This does indeed appear to be the case.

**Comment by behlendorf** [comment]
@adamdmoss thanks for running those tests.  I believe your results confirm what we've observed in the past.  And why we opted to set a relatively small dnode limit (since it keeps them from monopolizing the ARC).

The main reason these default tunings keep coming up is the high CPU usage they can be caused by metadata intensive workloads (find, rsync, etc).  That behavior I think we may be able to reasonably improve with a slightly more conservative version of this PR which just:

1) increases the `zfs_arc_meta_limit_percent` to 100%,
2) increases the `zfs_arc_dnode_limit_percent` to 20%, in order to keep _most_ workloads under the limit,
3) verify that limiting the system to one arc_prune task does reduce the CPU usage by avoiding contention.

Making further more substantial improvements would be good.  But this might be a reasonable middle group which covers most workloads.

**Comment by adamdmoss** [comment]
>     increases the zfs_arc_meta_limit_percent to 100%,
>     increases the zfs_arc_dnode_limit_percent to 20%, in order to keep most workloads under the limit,
>     verify that limiting the system to one arc_prune task does reduce the CPU usage by avoiding contention.

I gave that a spin:

```
(master)                                                                (PR w/20% dnode limit)
Starting tests...							Starting tests...
Big metadata crawl:	1:36.43 real,	10.86 user,	46.02 sys	Big metadata crawl:	1:39.12 real,	10.98 user,	46.58 sys
Big metadata crawl:	1:18.05 real,	11.63 user,	29.50 sys	Big metadata crawl:	1:13.66 real,	11.00 user,	28.67 sys
Little metadata crawl:	0:16.70 real,	0.44 user,	7.71 sys	Little metadata crawl:	0:14.66 real,	0.57 user,	7.65 sys
Little metadata crawl:	0:01.88 real,	0.27 user,	1.51 sys	Little metadata crawl:	0:06.11 real,	0.46 user,	3.27 sys
Little metadata crawl:	0:01.91 real,	0.24 user,	1.57 sys	Little metadata crawl:	0:06.01 real,	0.47 user,	3.19 sys
Little metadata crawl:	0:01.90 real,	0.25 user,	1.55 sys	Little metadata crawl:	0:05.99 real,	0.46 user,	3.16 sys
Assorted 5GB data read:	0:05.25 real,	0.60 user,	1.29 sys	Assorted 5GB data read:	0:07.18 real,	0.58 user,	1.62 sys
Assorted 5GB data read:	0:04.90 real,	0.53 user,	1.39 sys	Assorted 5GB data read:	0:07.10 real,	0.57 user,	1.54 sys
Assorted 5GB data read:	0:04.90 real,	0.58 user,	1.42 sys	Assorted 5GB data read:	0:07.12 real,	0.59 user,	1.56 sys
Assorted 5GB data read:	0:04.88 real,	0.53 user,	1.39 sys	Assorted 5GB data read:	0:07.11 real,	0.55 user,	1.61 sys
Assorted 5GB data read:	0:04.90 real,	0.56 user,	1.43 sys	Assorted 5GB data read:	0:07.14 real,	0.54 user,	1.61 sys
DROPPED CACHES...							DROPPED CACHES...
Little metadata crawl:	0:09.03 real,	0.35 user,	4.28 sys	Little metadata crawl:	0:06.18 real,	0.27 user,	3.96 sys
Little metadata crawl:	0:01.52 real,	0.20 user,	1.35 sys	Little metadata crawl:	0:01.53 real,	0.23 user,	1.31 sys
Little metadata crawl:	0:01.52 real,	0.24 user,	1.31 sys	Little metadata crawl:	0:01.52 real,	0.22 user,	1.34 sys
Little metadata crawl:	0:01.51 real,	0.24 user,	1.29 sys	Little metadata crawl:	0:01.53 real,	0.24 user,	1.30 sys
Assorted 5GB data read:	0:05.18 real,	0.55 user,	1.46 sys	Assorted 5GB data read:	0:05.16 real,	0.54 user,	1.44 sys
Assorted 5GB data read:	0:00.57 real,	0.27 user,	1.45 sys	Assorted 5GB data read:	0:00.56 real,	0.28 user,	1.43 sys
Assorted 5GB data read:	0:00.56 real,	0.29 user,	1.43 sys	Assorted 5GB data read:	0:00.56 real,	0.34 user,	1.36 sys
Assorted 5GB data read:	0:00.58 real,	0.31 user,	1.36 sys	Assorted 5GB data read:	0:00.56 real,	0.25 user,	1.48 sys
Assorted 5GB data read:	0:00.56 real,	0.30 user,	1.42 sys	Assorted 5GB data read:	0:00.56 real,	0.27 user,	1.40 sys
Finished tests.								Finished tests.
```

Raw data for the 20% dnode run:
[new3-PR20.txt](https://github.com/openzfs/zfs/files/5961746/new3-PR20.txt)

TL;DR: not an apparent improvement over the current state of the PR, overall probably a regression, though I wasn't able to verify whether it helps the CPU pegging from #9966

**Comment by adamdmoss** [comment]
Do I understand correctly that ARC pins metadata, and metadata pins dnodes?  Or is it another way around?
